src/picongpu/include/plugins/particleMerging/ParticleMerger.hpp

```diff
@@ -0,0 +1,253 @@
+            particles->fillAllGaps();
+        }
+
+
+        void setMappingDescription(MappingDesc* cellDescription)
+        {
+            this->cellDescription = cellDescription;
+        }
+
+
+        void pluginRegisterHelp(po::options_description& desc)
+        {
+            desc.add_options()
+            (
+                ( this->prefix + ".period" ).c_str(),
+                po::value< uint32_t > (
+                    &this->notifyPeriod
+                )->default_value( 0 ),
+                "enable plugin [for each n-th step]"
+            )
+            (
+                ( this->prefix + ".minParticlesToMerge" ).c_str(),
+                po::value< uint32_t > (
+                    &this->minParticlesToMerge
+                )->default_value( 8 ),
+                "minimal number of macroparticles needed to merge"
+                " the macroparticle collection into a single macroparticle."
+            )
+            (
+                ( this->prefix + ".posSpreadThreshold" ).c_str(),
+                po::value< float_X > (
+                    &this->posSpreadThreshold
+                )->default_value( 0.5 ),
+                "Below this threshold of spread in position macroparticles"
+                " can be merged [unit: cell edge length]."
+            )
+            (
+                ( this->prefix + ".absMomSpreadThreshold" ).c_str(),
+                po::value< float_X > (
+                    &this->absMomSpreadThreshold_mc
+                )->default_value( -1.0 ),
+                "Below this absolute threshold of spread in momentum"
+                " macroparticles can be merged [unit: m_el * c]."
+                " Disabled for -1 (default)."
+            )
+            (
+                ( this->prefix + ".relMomSpreadThreshold" ).c_str(),
+                po::value< float_X > (
+                    &this->relMomSpreadThreshold
+                )->default_value( -1.0 ),
+                "Below this relative (to mean momentum) threshold of spread in"
+                " momentum macroparticles can be merged [unit: none]."
+                " Disabled for -1 (default)."
+            )
+            (
+                ( this->prefix + ".minMeanEnergy" ).c_str(),
+                po::value< float_64 > (
+                    &this->minMeanEnergy_keV
+                )->default_value( 511.0 ),
+                "minimal mean kinetic energy needed to merge the macroparticle"
+                " collection into a single macroparticle [unit: keV]."
+            );
+        }
+
+        std::string pluginGetName() const
+        {
+            return this->name;
+        }
+
+    private:
+
+        void pluginLoad()
+        {
+            if( notifyPeriod == 0 )
+                return;
+
+            Environment<>::get().PluginConnector().setNotificationPeriod(
+                this,
+                notifyPeriod
+            );
+
+            // clean user parameters
+            PMACC_ASSERT_MSG(
```

Comment:
please use `PMACC_VERIFY_MSG` for all checks, to avoid that this checks are disabled when compiling in release mode.